### DISCUSSION TITLE
TypeScript: Am I using `MachineConfig` correctly?

### DISCUSSION THREAD

**Discussion by mackenco** [statelyai/xstate#2302] [Q&A]
I'm typing a machine using a [config object](https://xstate.js.org/docs/guides/typescript.html#config-objects):
```ts
interface LightStateSchema {
  states: {
    green: {};
    yellow: {};
  };
}

interface LightEvent { type: 'TIMER' }

interface LightContext { elapsed: number }

const lightMachine = MachineConfig<LightContext, LightStateSchema, LightEvent> = {
  id: 'lightMachine',
  states: {
    green: {},
  }
}

lightMachine.states.green = {
  on: {
    TIMER: { target: 'yellow' }
}
```
(this is a somewhat contrived example, but matches what I'm doing in the "real world"). `MachineConfig` doesn't define an `id` or `states`, so this returns errors when accessing `lightMachine.id` or `lightMachine.states.green` that either might be undefined. We can fix the latter with:
```ts
lightMachines.states!.green = ...
```
but this feels wrong? Right now I'm getting around this by extending `MachineConfig` to more closely match the `StateMachine` interface (what's returned from `createMachine`):
```ts
interface MachineConfig<Context, Schema extends StateSchema, Event extends EventObject>
  extends XStateMachineConfig<Context, Schema, Event> {
  id: string;
  states: StatesConfig<Context, Schema, Event>;
}
```

Any other ideas? Is the type of `MachineConfig` "correct"?

**Comment by Andarist** [comment]
Could you prepare a codesandbox of the problem that you are facing? `MachineConfig` extends `StateNodeConfig`:
https://github.com/davidkpiano/xstate/blob/6c8533846d659c5a633c3050aab120d03dbe6e07/packages/core/src/types.ts#L673-L677
and that type defines those:
https://github.com/davidkpiano/xstate/blob/6c8533846d659c5a633c3050aab120d03dbe6e07/packages/core/src/types.ts#L493

**Comment by mackenco** [comment]
[Codesandbox here](https://codesandbox.io/s/competent-dewdney-j55u0?file=/src/index.ts). You'll see that TS errors because `lightMachine.states` and `lightMachine.id` are both possibly `undefined`. I understand both of those are optional on `MachineConfig`, but the properties are present on the object itself. Thanks!

**Reply by davidkpiano** [reply]
Yeah, the typings are a bit loose for these (we can tighten them), but the quick workaround is to expect them to exist:

```ts

lightMachine.states!.green = {
  on: {
    TIMER: { target: "yellow" },
    POWER_OUTAGE: { target: "red" }
  }
};
```

**Reply by mackenco** [reply]
Makes sense - thanks for the reply!

**Reply by Andarist** [reply]
Actually - the types are correct here, both of those properties are actually optional for the `MachineConfig`. However, in general, using types like this might be more cumbersome than it is worth it for the consumer. Thanks to the structural typing nature of TypeScript you don't actually have to annotate this like this and you can just do stuff like this:
```ts
const lightMachine = {} as const
```
and for the most part the result should be accepted wherever `MachineConfig` is required.
